src/Access/LDAPAccessStorage.cpp

```diff
@@ -0,0 +1,265 @@
+        throw Exception("Empty 'server' field for LDAP user directory.", ErrorCodes::BAD_ARGUMENTS);
+
+    std::set<String> roles_cfg;
+    if (has_roles)
+    {
+        Poco::Util::AbstractConfiguration::Keys role_names;
+        config.keys(prefix_str + "roles", role_names);
+
+        // Currently, we only extract names of roles from the section names and assign them directly and unconditionally.
+        roles_cfg.insert(role_names.begin(), role_names.end());
+    }
+
+    access_control_manager = access_control_manager_;
+    ldap_server = ldap_server_cfg;
+    default_role_names.swap(roles_cfg);
+    roles_of_interest.clear();
+    role_change_subscription = access_control_manager->subscribeForChanges<Role>(
+        [this] (const UUID & id, const AccessEntityPtr & entity)
+        {
+            return this->processRoleChange(id, entity);
+        }
+    );
+}
+
+
+bool LDAPAccessStorage::isConfiguredNoLock() const
+{
+    return !ldap_server.empty() &&/* !roles.empty() &&*/ access_control_manager;
+}
+
+
+void LDAPAccessStorage::processRoleChange(const UUID & id, const AccessEntityPtr & entity)
+{
+    auto role_ptr = typeid_cast<std::shared_ptr<const Role>>(entity);
+    if (role_ptr)
+    {
+        if (default_role_names.find(role_ptr->getName()) != default_role_names.end())
+        {
+            auto update_func = [&id](const AccessEntityPtr & cached_entity) -> AccessEntityPtr
+            {
+                auto user_ptr = typeid_cast<std::shared_ptr<const User>>(cached_entity);
+                if (user_ptr && user_ptr->granted_roles.roles.find(id) == user_ptr->granted_roles.roles.end())
+                {
+                    auto clone = user_ptr->clone();
+                    auto user_clone_ptr = typeid_cast<std::shared_ptr<User>>(clone);
+                    user_clone_ptr->granted_roles.grant(id);
+                    return user_clone_ptr;
+                }
+                return cached_entity;
+            };
+
+            memory_storage.update(memory_storage.findAll<User>(), update_func);
+            roles_of_interest.insert(id);
+        }
+    }
+    else
+    {
+        if (roles_of_interest.find(id) != roles_of_interest.end())
+        {
+            auto update_func = [&id](const AccessEntityPtr & cached_entity) -> AccessEntityPtr
+            {
+                auto user_ptr = typeid_cast<std::shared_ptr<const User>>(cached_entity);
+                if (user_ptr && user_ptr->granted_roles.roles.find(id) != user_ptr->granted_roles.roles.end())
+                {
+                    auto clone = user_ptr->clone();
+                    auto user_clone_ptr = typeid_cast<std::shared_ptr<User>>(clone);
+                    user_clone_ptr->granted_roles.revoke(id);
+                    return user_clone_ptr;
+                }
+                return cached_entity;
+            };
+
+            memory_storage.update(memory_storage.findAll<User>(), update_func);
+            roles_of_interest.erase(id);
```

Comment:
Actually class `GrantedRoles` can be used outside the class `User`. If you change the type of `roles_of_interest` to `GrantedRoles` the code will become shorter cause you will able to use a simple assignment `user_clone_ptr->granted_roles = roles_of_interests` instead.

Reply:
I believe, that `GrantedRoles` object assignment will be heavier that erasing a single element. (And we have to copy the member object during cloning anyway.)

> And it seems the things will be better if you make the updating in two separate stages:

Not sure I understood that one.
Technically, I do it in two steps:
1. when a role is being added and its name is in my `default_role_names`
2. when a role is being removed and its id is in my `roles_of_interest`

`roles_of_interests` is being updated in both these scenarios.

Reply:
No, I meant something like this:

```
class LDAPAccessStorage
{
    std::set<String> default_role_names;
    std::unordered_map<UUID, String> roles_of_interest;
};


void LDAPAccessStorage::setConfiguration()
{
    ...

    roles_of_interest.clear();
    role_change_subscription = access_control_manager->subscribeForChanges<Role>(
        [this] (const UUID & id, const AccessEntityPtr & entity)
        {
            return this->processRoleChange(id, entity);
        }
    );
    for (const String & role_name : default_role_names)
    {
        if (auto role_id = access_control_manager->find<Role>(role_name))
            roles_of_interest.insert(*role_id);
    }
}


void LDAPAccessStorage::processRoleChange(const UUID & id, const AccessEntityPtr & entity)
{
    std::scoped_lock lock(mutex);

    /// Update `roles_of_interests`.
    auto role = typeid_cast<std::shared_ptr<const Role>>(entity);
    bool need_to_update_users = false;
    if (role && default_role_names.count(role->getName()))
    {
        /// If a role was created with one of the `default_role_names` or renamed to one of the `default_role_names`,
        /// then set `need_to_update_users`.
        need_to_update_users = roles_of_interest.insert_or_assign(id, role->getName()).second;
    }
    else
    {
        /// If a role was removed or renamed to a name which isn't contained in the `default_role_names`,
        /// then set `need_to_update_users`.
        need_to_update_users = roles_of_interest.erase(id) > 0;
    }

    /// Update users which have been created.
    if (need_to_update_users)
    {
        auto update_func = [&id](const AccessEntityPtr & entity) -> AccessEntityPtr
        {
            if (auto user = typeid_cast<std::shared_ptr<const User>>(entity))
            {
                auto changed_user = typeid_cast<std::shared_ptr<User>>(user->clone());
                auto & granted_roles = changed_user->granted_roles.roles;
                granted_roles.clear();
                boost::range::copy(roles_of_interest | boost::adaptors::map_keys, std::inserter(granted_roles));
                return changed_user;
            }
            return entity;
        };
        memory_storage.update(memory_storage.findAll<User>(), update_func);
    }
}
```

I think this is a better way because `roles_of_interests` is kept correct all the time, and updating of the created users is performed only if it's necessary. And also it works fine in case when some of the roles are renamed.

Reply:
Pushed the proposed changes with slight modifications.